blog-post/blog-post.md

```diff
@@ -0,0 +1,739 @@
+associated code generation pipeline down to Yul. In order to avoid getting bogged down in
+bikeshedding around syntax, and with the desire to validate our core ideas with a working
+implementation as soon as possible, we moved ahead with a provisional syntax. You can expect
+extensive changes before release. Our current intention is to eventually match the syntax of Classic
+Solidity as much as possible. For any new syntax, you should expect that the final version will feel
+closer to languages like TypeScript or Rust than it does right now.
+
+## New Language Features
+
+Core Solidity takes ideas from pure functional programming languages (e.g. Haskell, Lean), as well
+as modern systems languages (e.g. Rust, Zig). We are extending Solidity with the following new
+features:
+
+- Algebraic datatypes (a.k.a sum / product types) and pattern matching
+- Generics / parametric polymorphism
+- Traits / typeclasses
+- Type inference
+- Higher order and anonymous functions
+- Compile time evaluation
+
+We think that taken together these core primitives will enable developers to produce stronger
+abstractions, write more modular and reusable code, and leverage the type system to enforce core
+safety properties.
+
+We will continue to support the kind of low level access to the EVM that is often required by
+production implementations: assembly will remain a core primitive, and we will extend it with the
+ability to directly call functions from the high level language in assembly blocks. Users will be
+able to disable the languages built in abstractions (e.g. contract dispatch generation, ABI
+decoding, default storage layout generation), following the "pay for what you use" philosophy of
+languages like Rust and C++.
+
+### Algebraic data types and pattern matching
+
+Algebraic Data Types provide a principled foundation for data modeling through the composition of
+sum and product types. Sum types represent exclusive alternatives: a value inhabits exactly one
+variant. Product types combine multiple values into structured tuples. The combination of these two
+primitives allows for the construction of precise types that can make invalid states completely
+unrepresentable, allowing the type system to guarantee core program invariants entirely at
+compile time.
+
+Lets start with a very simple type:
+
+```solidity
+data Bool = True | False
+```
+
+Types are sets of values. The left hand side of the above statement defines the name of a new type
+(`Bool`), and the right hand side defines the set of allowed values of the `Bool` type (`True` or
+`False`). The `|` lets us introduce new alternative value constructors.
+
+Alternatives can also hold values. This lets us implement the same kind of patterns as [User Defined
+Value Types](https://docs.soliditylang.org/en/latest/types.html#user-defined-value-types) would in
+Classic Solidity. We can consider a simple 18 decimal fixed point (a `wad`):
+
+```solidity
+data wad = wad(uint256)
```

Comment:
Note for those unfamiliar that the `wad` name is fairly common for an 18 decimal fixed point and is used by both [solady](https://github.com/Vectorized/solady/blob/main/src/utils/FixedPointMathLib.sol) and [solmate](https://github.com/transmissions11/solmate/blob/main/src/utils/FixedPointMathLib.sol). We can switch to a more descriptive ABDK / prbmath style naming if people have strong opinions, but I personally find `wad` a little cleaner and less cluttered.